[IMP] test_mail_full: add portal flow test

pyduf opened the pull request:
The tests consist in sending a mail related to a record to a customer and
checking that the record can be viewed through the embedded link:
- either in the backend if the user is connected and has the right to
- or in the portal otherwise

Task-2797311

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/105405).

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+    The tests consist in sending a mail related to a record to a customer and checking that the record can be viewed
+    through the embedded link:
+    - either in the backend if the user is connected and has the right to
+    - or in the portal otherwise """
```
```suggestion
    - or in the portal otherwise
    """
```

pyduf replied:
Thanks, it is updated.

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+
+@tagged('portal')
+class TestPortalFlow(MailCommon, HttpCase):
+    """Portal flow test: share a link by email to a customer without an account for viewing a record through the portal.
```
This is already in the class name
```suggestion
    """Share a link by email to a customer without an account for viewing a record through the portal.
```

pyduf replied:
Yes indeed, thanks.

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+    def test_employee_access(self):
+        """Check that the access link redirects an employee to the backend for viewing the record."""
+        self.authenticate(self.env.user.login, self.env.user.login)
+        res = self.url_open(self._get_record_access_url())
```
can we test also that if the token is wrong, we still can access the record ? (because we are logged)

pyduf replied:
Yes, good idea, I have added the test.

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+    def test_customer_access_logged_without_access(self):
+        """Check that the link redirects the customer (without backend access) to the portal for viewing the record."""
+        self.authenticate(self.env.user.login, self.env.user.login)
+        res = self.url_open(self._get_record_access_url())
```
same here about wrong token

pyduf replied:
Yes indeed, some test were missing and actually not straightforward.

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+    def test_customer_access_not_logged(self):
+        """Check that the access link redirects the customer (not logged) to the portal for viewing the record."""
+        res = self.url_open(self._get_record_access_url())
+        self.assertEqual(res.status_code, 404)  # No controller implemented for the test
```
having a 404 error here because the route is not implemented is a bit mhe, can we implement the route ? (so in the test we will have a 200 code, which is more explicit)

pyduf replied:
Yes that’s better to test the whole flow, thanks.

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+        """Check that the access link redirects the customer (not logged) to the portal for viewing the record."""
+        res = self.url_open(self._get_record_access_url())
+        self.assertEqual(res.status_code, 404)  # No controller implemented for the test
+        self.assertIn(f'/my/test_portal/{self.record_customer.id}', res.url)
```
can we test that if the token is wrong, we get an error ?

pyduf replied:
Yes, thanks. I have added the test but for info, the current implementation redirect to discuss in that case.

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+            composer._action_send_mail()
+
+        self.assertEqual(len(self._mails), 1)
+        self.assertIn(self._get_record_access_url(), self._mails[0].get('body').replace('&amp;', '&'))
```
can we html escape `_get_record_access_url` instead of `.replace('&amp;', '&')` ?
(same above)

pyduf replied:
Yes good idea, thanks

std-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +163,111 @@ def test_portal_mixin(self):
+
+        Other tests below check that that same link has the correct behavior.
+        This test follows the common use case by using a template while the next send the mail without a template."""
+        composer = self.env['mail.compose.message'].with_context(self._get_composer_context(self.mail_template.id))\
```
not a big fan of `\`

instead, can `_get_composer_context` return the model directly ? (`self.env['mail.compose.message'].with_context(context)`)

pyduf replied:
Yes good idea, thanks

std-odoo reviewed (commented):
@pyduf Hi :) 

I have some comments, thanks for your work :)

pyduf commented:
> @pyduf Hi :)
> 
> I have some comments, thanks for your work :)

Thanks for the review. There were indeed some test missing. 
For info, I have also removed the _get_record_access_url and added a class variable instead:  record_access_url (idem for newly added record_access_url_wrong_token). I test those in test_initial_data method.

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+            'default_use_template': template_id,
+            'default_template_id': template_id,
+            'default_composition_mode': 'comment',
+            'mark_so_as_sent': True,
```
Probably not necessary :D (copy paste maybe ^^ ?)

pyduf replied:
Yes indeed, I have removed the last line (about sale order), thanks.

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+            'mark_so_as_sent': True,
+            'default_email_layout_xmlid': 'mail.mail_notification_layout_with_responsible_signature',
+            'force_email': True,
+            'model_description': 'Quotation',
```
Normally not necessary

pyduf replied:
Yes it works without, thanks.

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+
+    def test_initial_data(self):
+        """Test some initial value."""
+        self.assertNotEqual(self.record_access_url, self.record_access_url_wrong_token)
```
Could be great to parse the url and check the URL itself (containing /my/test_portal/rec.id + token (and maybe pid and hash, depending on how we call things, don't remember all details of the call stack ^^).

pyduf replied:
Yes, indeed, that’s better to do a more in-depth validation of init data.

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+        Other tests below check that that same link has the correct behavior.
+        This test follows the common use case by using a template while the next send the mail without a template."""
+        composer = self._get_composer_with_context(self.mail_template.id).create({})
+        # currently onchange necessary
```
I know this comment :disguised_face:

pyduf replied:
:-) I have copied it in case we want to change that in the future.

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+
+
+@tagged('portal')
+class TestPortalFlow(MailCommon, HttpCase):
```
I would define it before TestPortalMixin as classes are called by alphabetical order (easier when doing refactoring, having tests following the file order, better when having several tests to fix or check :D )

pyduf replied:
Changed.

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+        self.assertIn(html_escape(self.record_access_url), self._mails[0].get('body'))
+        self.assertEqual('Quotation', self._mails[0].get('subject'))  # Check that the template is not used
+
+    def test_customer_access_not_logged(self):
```
Side note: could you order tests alphabetically in a given class ? Same reason as for classes: easier when doing refactoring, having tests following the file order, better when having several tests to fix or check :D

pyduf replied:
Done, the only test that I didn’t move is test_initial_data because I think it should be the first test. What do you think?

tde-banana-odoo replied:
That's ok :) we should find a synonym beginning with a :thinking:

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+        res = self.url_open(self.record_access_url_wrong_token)
+        self.assertEqual(res.status_code, 200)
+        self.assertIn('/web/login?redirect=#', res.url)
+        self.assertIn('model=mail.test.portal', res.url)
```
Note that we could parse the url and check the fragments

pyduf replied:
Indeed, it is more reliable. I have replaced all URL check by checking parsed parts.

tde-banana-odoo reviewed (commented):
Zboing! Thanks for the PR, glad to see we are finally adding some tests for portal tokens :D added some additional comments while reading :)

pyduf commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.customer = cls.env['res.partner'].create({
+            'name': 'Wood Corner',
```
Good idea, it is added.

pyduf commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+            'name': 'Wood Corner',
+            'email': 'wood_corner@example.com',
+        })
+        cls.record_customer = cls.env['mail.test.portal'].create({
```
Yes indeed, it is changed.

pyduf commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+        super().setUpClass()
+        cls.customer = cls.env['res.partner'].create({
+            'name': 'Wood Corner',
+            'email': 'wood_corner@example.com',
```
Yes, it is done.

pyduf commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+        })
+        cls.mail_template = cls.env['mail.template'].create({
+            'name': 'Test send quotation',
+            'subject': 'Your quotation',
```
Yes, indeed, it is changed.

pyduf commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+            'auto_delete': True,
+            'email_from': '{{ (object.user_id.email_formatted or user.email_formatted) }}',
+            'partner_to': '{{ object.partner_id.id }}',
+            'model_id': cls.env.ref('test_mail_full.model_mail_test_portal').id,
```
I have added the lang. Do you want me to add test about translation ?

tde-banana-odoo replied:
Well would say maybe later, so that was already land with this improved tests :)

tde-banana-odoo commented:
@pyduf beware ci style is red :)

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -145,6 +147,188 @@ def test_portal_share_comment(self):
+            'country_id': cls.env.ref('base.be').id,
+            'email': 'mdelvaux34@example.com',
+            'lang': 'en_US',
+            'mobile': '+32492131422',
```
Just asking, is it a fake number ? I know there are some reserved working numbers for testing (not remembering them, maybe internet could answer this ^^ ) .

pyduf replied:
Yes good remark. I didn’t know it existed but it is logical. I have found on wikipedia a range of number for France of “fictitious number” so I have updated the number to use one of them. (I had used a generator but I don’t know if the number is in use or not.)

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -145,6 +147,188 @@ def test_portal_share_comment(self):
+                        sort=True),
+             parsed_url[4]))
+
+    def _assert_URL(self, url, path, expected_fragment_params=None, expected_in_query=None):
```
Would add a small docstring :) and maybe make it public (aka without underscore) ? Not sure we have to keep it private-like.

pyduf replied:
Yes, good idea. By doing docstring helped me to make an easier method to understand.

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -145,6 +147,188 @@ def test_portal_share_comment(self):
+            for key, value in expected_fragment_params.items():
+                self.assertEqual(fragment_params.get(key), value, f'Param: {key}')
+        if expected_in_query:
+            self.assertIn(expected_in_query, parsed_url.query, f'Query of {path}')
```
If you add a message, I would add the expected versus actual in it, to ease understand the fail (same for above) otherwise you still have to add a print when it fails :D

pyduf replied:
he idea was to give the context in the message but yes, it is better to set the expected value as well as the context. Here is how it displays now: 

AssertionError: 'mail.test.portal' != 'mail.test.portal-'
- mail.test.portal
+ mail.test.portal-
?                 +
 : Expected: "model=mail.test.portal" (for path: /web/login)

tde-banana-odoo commented on addons/test_mail_full/tests/test_portal.py:
```diff
@@ -162,3 +165,143 @@ def test_portal_mixin(self):
+        })
+        cls._create_portal_user()
+        # Get access URL that is embedded in the customer email for accessing the record.
+        for group_name, group_func, group_data in cls.record_customer.sudo()._notify_get_recipients_groups():
```
Well ok. Maybe we could think of extracting the url generation code and improve it a bit in the future, and add tests for pid and hash, but will be for a next iteration :) here we just add the first steps :)

tde-banana-odoo reviewed (commented):
Zboing ! Thanks for update, few comments, check ci style then we are good to go I think :)

pyduf commented:
> @pyduf beware ci style is red :)

Thanks for the review. 
I saw it but was uncertain to solve it or not because it is legitimate but as we don't check "hash", I have removed that parameter that was causing the style problem.

Details: The ci style was due to the “hash” parameter in the controller (which redefine locally a built-in function). As we don’t care about “pid and hash” parameters here, I have used **kwargs in the controller for them. So I get rid of that warning and don’t get warning for ignored parameters.

tde-banana-odoo commented:
Ah yes indeed, 'hash' is the portal standard naming... we can keep it if you prefer, I can override ci/style when it is legit.

tde-banana-odoo commented:
@robodoo r+
